Changed Range defs, Added Discontinutes, Added Poisson Distribution, Tested Everything

toast682 opened the pull request.

vincenzocoia commented on R/mix.R:
```diff
@@ -108,142 +113,176 @@ is.mix <- function(object) inherits(object, "mix")
+  for (i in 1:length(distributions)) {
+    distributions[[i]]$size <- distributions[[i]]$size *
+      object$components$probs[i]
+    res <- rbind(res, distributions[[i]])
```
`rbind` can be quite slow when run many times like this. I have two ideas to tackle this:

1. Move the `rbind` call outside of the `for` loop. Once the `for` loop finishes, you can bind all the data frames at once with `do.call(rbind, object$components)`. OR:
2. You could parse out the `size` and `location` columns into a list of these vectors: `sizes <- lapply(distributions, `[`, "size")` and `locations <- lapply(distributions, `[`, "location")`. Go ahead and modify the size vectors as you're doing currently in the `for` loop (such as `sizes[[i]] <- sizes[[i]] * object$components$probs[i]`). Then, flatten the lists as a single vector: `size_vec <- c(sizes, recursive = TRUE)` and `location_vec <- c(locations, recursive = TRUE)`. Then feed the size and location vectors into `aggregate_weights()`. This method sounds like more work, but I suspect the function would run quickest this way (although I'm not 100% sure).

vincenzocoia reviewed (commented):
Hi @toast682 , this is looking really good. I'm not finished going through everything yet -- stopping at `R/representation-cdf.R` -- but this should be a good start. I'll get around to the rest of the code soon. Well done!!!

vincenzocoia commented on tests/testthat/test-discontinuities.R:
```diff
@@ -0,0 +1,284 @@
+  )
+  expect_equal(
+    discontinuities(dst_finite(1:5, rep(0.2, 5)), 58, 80),
+    tibble::tibble(
```
Would this not be `make_empty_dicontinuities_df()`?

vincenzocoia commented on tests/testthat/test-discontinuities.R:
```diff
@@ -0,0 +1,284 @@
+test_that("Discontinuites Degenerate works", {
+  expect_equal(
+    discontinuities(dst_degenerate(1), -Inf, Inf),
+    tibble::tibble(location = c(1), size = c(1))
```
No need for `c()` when there's only one entry in the vector. So, `c(1)` should be `1`, etc. -- applies for downstream lines, too.

vincenzocoia commented on tests/testthat/test-discontinuities.R:
```diff
@@ -0,0 +1,284 @@
+      dst_norm(1, 4),
+      dst_norm(1, 3)
+    ), 1, 3),
+    tibble::tibble(location = numeric(), size = numeric())
```
`make_empty_...()`?

vincenzocoia commented on tests/testthat/test-discontinuities.R:
```diff
@@ -0,0 +1,284 @@
+      dst_norm(1, 4),
+      dst_unif(1, 3)
+    ), 4, 5),
+    tibble::tibble(location = numeric(), size = numeric())
```
`make_empty_...()`?

vincenzocoia commented on tests/testthat/test-discontinuities.R:
```diff
@@ -0,0 +1,284 @@
+      dst_norm(1, 4),
+      dst_gpd(1, 3, 34)
+    ), 7, 2342342),
+    tibble::tibble(location = numeric(), size = numeric())
```
`make_empty_...()`?

vincenzocoia commented on tests/testthat/test-discontinuities.R:
```diff
@@ -0,0 +1,284 @@
+        dst_norm(1, 5)
+      ), -94, 5
+    ),
+    tibble::tibble(
```
`make_empty_...()`?

vincenzocoia commented on tests/testthat/test-dst_pois.R:
```diff
@@ -0,0 +1,36 @@
+  .dst <- dst_pois(lambda = lambda)
+  expect_identical(mean(.dst), lambda)
+  expect_identical(variance(.dst), lambda)
+  expect_identical(distplyr::sd(.dst), sqrt(lambda))
```
Hmm, I don't remember where I left off with the `sd()` function. It looks like the regular `sd()` function is not an S3 generic function, which means that we can't make specific methods like `sd.dst()`. And, it's not a good idea to override base R's `sd()` function when distplyr is loaded. 

I think I remember encountering the same problem with `var()` for variance. I ended up making a generic called `variance()` instead.

@toast682 If distplyr defines an `sd()` function, we should probably change it to something else, like I did for the variance. Can you do this? What do you think of `stdev()` as the generic function? I think `standard_deviation()` might be too verbose. As for specific methods, I think we would only need a standard `stdev.dst()` that takes the square root of the variance -- no need to define special ones for the other distributions, I think.

vincenzocoia commented on tests/testthat/test-dst_pois.R:
```diff
@@ -0,0 +1,36 @@
+test_that("dst_pois basic functions work", {
```
These tests look great for now. I think I've come up with a more robust testing framework that doesn't rely on formulas (there's just too much room for human error -- for example, if the formula we're referring to has a mistake, or is talking about something different). We don't have to do anything about this now, but at some point, it would be good to check the mean, variance, etc. against calculations of these quantities using things like density/cdf/etc.

Just something to keep in mind -- nothing to do about this right now.

vincenzocoia commented on tests/testthat/test-dst_pois.R:
```diff
@@ -0,0 +1,36 @@
+test_that("dst_pois pmf and cdf functions works", {
+  lambda <- 2
+  .dst <- dst_pois(lambda = lambda)
+  cdf_one <- stats::dpois(0, lambda) + stats::dpois(1, lambda)
```
There's an easier way to calculate the cdf -- the `ppois()` function does this adding for us. So:

```
cdf_one <- stats::ppois(1, lambda)
cdf_two <- stats::ppois(2, lambda)
...
```

Even better, `ppois()` is vectorized, meaning that if you input a vector, `ppois()` will output a vector. So, instead of: 

```
expect_equal(eval_cdf(.dst, c(1, 2, 9)), c(cdf_one, cdf_two, cdf_nine))
```

we can just write:

```
expect_equal(eval_cdf(.dst, c(1, 2, 9)), stats::ppois(c(1, 2, 9), lambda))
```

vincenzocoia commented on tests/testthat/test-range.R:
```diff
@@ -0,0 +1,59 @@
+  )
+})
+
+test_that("range for empirical dst works", {
```
Flagging this chunk as being dependent on the "datasets" R package.

vincenzocoia commented on tests/testthat/test-sd.R:
```diff
@@ -2,6 +2,6 @@ x <- c(4, 5, 6, 7, 80)
 y <- c(TRUE, FALSE)
 
 test_that("stats::sd is called on non-distributions", {
```
I think I remember what I did with `sd()`: I made an `if` statement that dispatched base R's `sd()` if necessary. This is a shotty way of doing things, and we should stick to the `stdev()` generic (or whatever we decide) instead. 

Anyway, once our new generic for standard deviation is created, this test `test_that("stats::sd is called on non-distributions"` can be deleted.

vincenzocoia reviewed (commented):
@toast682 I'm finished reviewing this pull request, and have added a few more in-line comments. Please let me know if you have any questions!

toast682 commented:
Hey, I added all the changes you wanted. I removed the sd function completely and added another S3 generic method called stdev instead and changed all the sd generic methods to stdev. Ill talk more about the other changes in today's, meeting

vincenzocoia reviewed (approved):
Thank you!!!

vincenzocoia commented:
I'm making a few minor adjustments, and will merge shortly.
